File: ddd-application/src/main/java/victor/training/ddd/agile/service/ProductService.java
package victor.training.ddd.agile.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;
import victor.training.ddd.agile.entity.Product;
import victor.training.ddd.agile.dto.ProductDto;
import victor.training.ddd.agile.entity.ProductOwner;
import victor.training.ddd.agile.repo.ProductRepo;

@Slf4j
@RestController
@RequiredArgsConstructor
public class ProductService {
    private final ProductRepo productRepo;

    @PostMapping("products")
    public Long createProduct(@RequestBody ProductDto dto) {
        if (productRepo.existsByCode(dto.getCode())) {
            throw new IllegalArgumentException("Code already defined");
        }
        Product product = new Product(dto.getCode(), dto.getName())
                .setTeamMailingList(dto.getMailingList())
                .setProductOwner(new ProductOwner(dto.getPoEmail(), dto.getPoName(), dto.getPoPhone()));
        return productRepo.save(product).getId();
    }

    @GetMapping("products/{id}")
    public ProductDto getProduct(@PathVariable long id) {
        Product product = productRepo.findOneById(id);
        ProductDto dto = new ProductDto()
                .setId(product.getId())
                .setName(product.getName())
                .setCode(product.getCode())
                .setMailingList(product.getTeamMailingList())
                .setPoEmail(product.getProductOwner().getOwnerEmail())
                .setPoName(product.getProductOwner().getOwnerName())
                .setPoPhone(product.getProductOwner().getOwnerPhone())
                ;
        return dto;
    }
}

File: ddd-application/src/main/java/victor/training/ddd/agile/service/ProductService.kt
package victor.training.ddd.agile.service

import org.springframework.web.bind.annotation.*
import victor.training.ddd.agile.dto.ProductDto
import victor.training.ddd.agile.entity.Product
import victor.training.ddd.agile.repo.ProductRepo

@RestController
class ProductService(private val productRepo: ProductRepo) {

    @PostMapping("products")
    fun createProduct(@RequestBody dto: ProductDto): Long? {
        require(!productRepo.existsByCode(dto.code)) { "Code already defined" }
        val product = Product(dto.code, dto.name, dto.mailingList)
        return productRepo.save(product).id
    }

    @GetMapping("products/{id}")
    fun getProduct(@PathVariable id: Long): ProductDto {
        val product = productRepo.findOneById(id)
        return ProductDto(product.id, product.code, product.name, product.teamMailingList)
    }
}